tests/docs/test_complete_examples.py

```diff
@@ -0,0 +1,185 @@
+import csv
+import tempfile
+import typing
+import uuid
+import warnings
+from collections.abc import Iterator
+from dataclasses import dataclass, field
+from pathlib import Path
+
+import pytest
+
+import ehrql.main
+
+
+@dataclass
+class DatasetDefinitionExample:
+    """Store details of a dataset definition example."""
+
+    source_filename: str
+    start: int = None
+    end: int = None
+    content_lines: list[str] = field(default_factory=lambda: [])
+
+    def read(self) -> str:
+        return "".join(self.content_lines)
+
+
+def discover_markdown_paths() -> list[str]:
+    """Find Markdown files in the documentation source path."""
+    docs_path = Path(__file__).parents[2] / "docs"
+    # Note that we're currently not excluding any paths like the generated docs.
+    # Do we need to? I don't think so, because if examples aren't labelled as
+    # ```ehrql then they won't be checked.
+    markdown_paths = list(docs_path.glob("**/*.md"))
+    return markdown_paths
+
+
+def find_complete_ehrql_examples_in_markdown(
+    file: typing.TextIO,
+) -> Iterator[DatasetDefinitionExample] | None:
+    """Read a Markdown file and extracts code blocks labelled as ```ehrql.
+
+    Incomplete ehrQL dataset definitions should be labelled as ```python,
+    and not with ```ehrql."""
+    # There are three kinds of possible ehrql code blocks that we might see.
+    # 1. A complete and correctly formatted example like:
+    #
+    # ```ehrql
+    # …
+    # ```
+    #
+    # 2. An incorrectly formatted example where we see another open code block
+    #    before a closed one.
+    #
+    # ```ehrql
+    # …
+    # ```something
+    #
+    # 3. An incorrectly formatted example where the code block is not closed
+    #    before the end of file.
+    #
+    # ```ehrql
+    # …
+    # EOF
+    in_ehrql_block = False
+    example = None
+
+    for line_number, line in enumerate(file, start=1):
+        # Check if the line marks the start of a ehrQL code block?
+        if line.startswith("```ehrql") or line.startswith("``` ehrql"):
```

Comment:
thought: I'd started with just trying to get the content of interest out. We could instead pull out *all* code blocks using the backtick formatting, and then filter those for the syntax(es) we want at the end. That might be `ehrql-dataset` and `ehrql-measure`, potentially.